specify/specify7
# Display parentCOG on CO and COG form

Fixes #5428
Fixes #5437
Fixes #5450 

⚠️ **Note:** This PR affects database migrations. See [migration testing instructions](https://specify.github.io/testing/pull_requests.html#prs-tagged-with-label-migration).

This pr allow to visualize the parentCOG on CO and COG form + add save blocker when trying to add a cog to itself 

### Checklist

      and self-explanatory (or properly documented)

### Testing instructions

!To test this a few modifications in the xml are needed. See bellow. (the parentCog will be displayed as a COJO subview) 

!!! after testing, when copy pasting the testing instructions, please only keep the - [ ] verify... remove all the rest 

Create in CO 
- open CO form 
- save 
- create a new COG parent in the COJO subview
- save 
- [ ] verify you can see the new parentCOG 
- [ ] verify you can open the new COG with edit tool 
- [ ] verify you can see the CO in the COG when you open the record

Use existant in CO
- open CO form 
- save 
- search for a COG parent in the COJO subview
- save 
- [ ] verify you can see the parent COG 
- [ ] verify you can open the parent COG with edit tool 
- [ ] verify you can see the CO in the COG when you open the record

Create in COG 
- open COG form 
- save 
- create a new COG parent in the COJO subview
- save 
- [ ] verify you can see the new parentCOG 
- [ ] verify you can open the new COG with edit tool 
- [ ] verify you can see the CO in the COG when you open the record

Use existant in COG
- open COG form 
- save 
- search for a COG parent in the COJO subview
- save 
- [ ] verify you can see the parent COG 
- [ ] verify you can open the parent COG with edit tool 
- [ ] verify you can see the CO in the COG when you open the record

Delete parentCOG from the parent side for CO
- open CO form 
- save 
- search for a COG parent in the COJO subview or create new one
- save 
- open COG parent using the edit tool 
- delete child CO line 
- save parent cog form dialog 
- close 
- reload the page 
- [ ] verify the parent cog has been removed from the child form 
PS: it is not possible to delete a parent on the child form

Delete parentCOG from the parent side for COG
- open COG form 
- save 
- search for a COG parent in the COJO subview or create new one
- save 
- open COG parent using the edit tool 
- delete child CO line 
- save parent cog form dialog 
- close 
- reload the page 
- [ ] verify the parent cog has been removed from the child form 
PS: it is not possible to delete a parent on the child form

Save blocker for COG to itself using children
- open COG form 
- save 
- add a child cog using QB or search 
- search for the cog itself 
- add it as a child 
- [ ] verify you have a tool tip blocker saying that you cannot add a cog to itself 

Save blocker for COG to itself using parent
- open COG form 
- save 
- add a parent cog using QB or search in the cojo parent subview
- search for the cog itself 
- add it as a parent 
- [ ] verify you have a tool tip blocker saying that you cannot add a cog to itself 

Check for parent cog when adding child on parent 
- open COG form 
- save 
- add a child cog using QB or search or create
- add the child
- save the form 
- open the child 
- [ ] verify you can see the name of the parent cog 

To add in your CO and COG form xml: 

```
<row>
<cell type="subview" viewname="CollectionObjectGroupJoin_Sub" id="43" name="cojo" colspan="11" />
</row>
```

Create a new collectionobjectgroupjoin_sub view def 
```
<viewdef name="CollectionObjectGroupJoin_sub” class="edu.ku.brc.specify.datamodel.CollectionObjectGroupJoin" type="form" gettable="edu.ku.brc.af.ui.forms.DataGetterForObj" settable="edu.ku.brc.af.ui.forms.DataSetterForObj">
	<desc>The Collection Object Group Join Form</desc>
	<enableRules/>
	<columnDef>50px,2px,30px,5px,50px,2px,30px,5px,50px,2px,100px,0px,50px,2px,100px,p:g</columnDef>
	<rowDef auto="true" cell="p" sep="2px"/>
	<rows>
		<row>
			<cell type="panel" colspan="20">
				<rows>
					<row>
						<cell type="label" labelfor="1"/>
						<cell type="field" id="1" name="isPrimary" uitype="checkbox" default="false" readonly="true"/>
						<cell type="label" labelfor="2"/>
						<cell type="field" id="2" name="isSubstrate" uitype="checkbox" default="false" readonly="true"/>
						<cell type="label" labelfor="5"/>
						<cell type="field" id="5" name="parentCog" uitype="querycbx" colspan="3"/>
					</row>
				</rows>
			</cell>
		</row>
	</rows>
</viewdef>

```
![Screenshot 2024-12-09 at 10 26 59 AM](https://github.com/user-attachments/assets/fd86f1dd-37ff-4300-a777-17af00f0f8a6)

![Screenshot 2024-12-09 at 10 27 33 AM](https://github.com/user-attachments/assets/04c3d694-c320-4c37-ae55-05ce2b67ac8a)
